src/GovUk.Education.ExploreEducationStatistics.Admin/Services/Public.Data/DataSetService.cs

```diff
@@ -170,66 +184,56 @@ private async Task<DataSetViewModel> MapDataSet(DataSet dataSet, CancellationTok
         };
     }
 
-    private async Task<List<Guid>> GetDataSetPreviousReleaseIds(DataSet dataSet, CancellationToken cancellationToken)
-    {
-        var previousReleaseFileIds = dataSet
-            .Versions
-            .Select(version => version.ReleaseFileId)
-            .ToList();
-
-        return await GetReleaseIdsForReleaseFiles(
-            contentDbContext,
-            previousReleaseFileIds,
-            cancellationToken);
-    }
-
-    private static async Task<List<Guid>> GetReleaseIdsForReleaseFiles(
-        ContentDbContext contentDbContext,
-        List<Guid> releaseFileIds,
+    private async Task<Dictionary<Guid, Dictionary<Guid, ReleaseVersion>>> GetPreviousReleasesByVersionByDataSet(
+        IReadOnlyList<DataSet> dataSets, 
         CancellationToken cancellationToken)
     {
-        return await contentDbContext
+        var dataSetsByPreviousReleaseFileId = dataSets
+            .SelectMany(ds => ds.Versions)
+            .ToDictionary(dsv => dsv.ReleaseFileId, dsv => dsv.DataSet);
+
+        var releasesByReleaseFileId = await contentDbContext
             .ReleaseFiles
-            .Where(releaseFile => releaseFileIds.Contains(releaseFile.Id))
-            .Select(releaseFile => releaseFile.ReleaseVersion.ReleaseId)
-            .ToListAsync(cancellationToken);
+            .Include(rf => rf.ReleaseVersion)
+            .Where(releaseFile => dataSetsByPreviousReleaseFileId.Keys.Contains(releaseFile.Id))
+            .ToDictionaryAsync(
+                rf => rf.Id,
+                rf => rf.ReleaseVersion,
+                cancellationToken);
+
+        return releasesByReleaseFileId
```

Comment:
Would rename this to `releaseVersionsByReleaseFileId`

Reply:
good shout! done